<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [16]:
import pandas as pd
import psycopg2

In [65]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

In [202]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [19]:
query_3_1 = f'''select  count(*)
            from vacancies v
         '''

In [20]:
df = pd.read_sql_query(query_3_1, connection)
df

C:\Users\bogd_\AppData\Local\Temp\ipykernel_21948\3851344990.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [28]:
query_3_2 = f'''select  count(e.name) 
            from employers e
         '''

In [29]:
df = pd.read_sql_query(query_3_2, connection)
df

C:\Users\bogd_\AppData\Local\Temp\ipykernel_21948\3588252056.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [30]:
query_3_3 = f'''select  count(a.name) 
            from areas a
         '''

In [32]:
df = pd.read_sql_query(query_3_3, connection)
df

C:\Users\bogd_\AppData\Local\Temp\ipykernel_21948\2018224393.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [33]:
query_3_4 = f'''select  count(i.name) 
            from industries i
         '''

In [34]:
df = pd.read_sql_query(query_3_4, connection)
df

C:\Users\bogd_\AppData\Local\Temp\ipykernel_21948\2172479966.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


count
0    294

***

кол-во вакансий = 49197 , кол-во работадателей = 23501, кол-во регионов =1362 , кол-во сфер деятельности = 294 

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [43]:
query_4_1 = f'''SELECT 
                a.name area, 
                COUNT(*) cnt
            FROM areas a
            JOIN VACANCIES v on a.id = v.area_id 
            Group by a.name
            order by cnt desc
            limit 5
            '''

In [44]:
df = pd.read_sql_query(query_4_1, connection)
df

C:\Users\bogd_\AppData\Local\Temp\ipykernel_21948\347208320.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


area   cnt
0           Москва  5333
1  Санкт-Петербург  2851
2            Минск  2112
3      Новосибирск  2006
4           Алматы  1892

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [45]:
query_4_2 = f'''SELECT 
                count(v.id) 
            FROM vacancies v
            where v.salary_from is not null or v.salary_to is not null
            '''

In [46]:
df = pd.read_sql_query(query_4_2, connection)
df

C:\Users\bogd_\AppData\Local\Temp\ipykernel_21948\2168503709.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [47]:
query_4_3=f''' select 
               round(avg(v.salary_from)) avg_lower,
               round(avg(v.salary_to)) avg_upper
            from vacancies v 



'''

In [67]:
df = pd.read_sql_query(query_4_3, connection)
df

C:\Users\bogd_\AppData\Local\Temp\ipykernel_21948\2222372203.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


avg_lower  avg_upper
0    71065.0   110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [76]:
query_4_4=f''' select v.schedule,
                v.employment
            from vacancies v 
            group by v.schedule,v.employment
            order by count(v.schedule) desc,count(v.employment) desc



'''

In [77]:
df = pd.read_sql_query(query_4_4, connection)
df

C:\Users\bogd_\AppData\Local\Temp\ipykernel_21948\2981602351.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


schedule           employment
0        Полный день     Полная занятость
1   Удаленная работа     Полная занятость
2      Гибкий график     Полная занятость
3   Удаленная работа  Частичная занятость
4     Сменный график     Полная занятость
5        Полный день           Стажировка
6     Вахтовый метод     Полная занятость
7        Полный день  Частичная занятость
8      Гибкий график  Частичная занятость
9        Полный день     Проектная работа
10  Удаленная работа     Проектная работа
11     Гибкий график           Стажировка
12    Сменный график  Частичная занятость
13  Удаленная работа           Стажировка
14     Гибкий график     Проектная работа
15    Сменный график           Стажировка
16    Вахтовый метод     Проектная работа
17    Сменный график     Проектная работа

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [74]:
query_4_5=f''' select v.experience
            from vacancies v 
            group by v.experience
            order by count(*)
'''

In [75]:
df = pd.read_sql_query(query_4_5, connection)
df

C:\Users\bogd_\AppData\Local\Temp\ipykernel_21948\2601166264.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)


experience
0         Более 6 лет
1           Нет опыта
2       От 3 до 6 лет
3  От 1 года до 3 лет

***

In [ ]:
# выводы по детальному анализу вакансий

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [85]:
query_5_1 =f''' select e.name
            from employers e 
            join vacancies v on e.id=v.employer_id
            group by e.name
            order by count(*) desc
'''

In [86]:
df = pd.read_sql_query(query_5_1, connection)
df

C:\Users\bogd_\AppData\Local\Temp\ipykernel_21948\1792814825.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)


name
0                      Яндекс
1                  Ростелеком
2                    Тинькофф
3                        СБЕР
4               Газпром нефть
...                       ...
14761            им. Тьюринга
14762        LLC Novalab Tech
14763   Компания Мир упаковки
14764  Лифинцев Олег Иванович
14765      ТрансЭнергоХолдинг

[14766 rows x 1 columns]

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [120]:
query_5_2 =f''' select a.name areas_name,
                count(distinct (e.id)) emp_cnt,
                count(distinct(v.id)) vacancy_cnt
            from areas a 
            left join employers e on a.id=e.area
            left join vacancies v on v.area_id=a.id
            where v.id is null
            group by a.name
            order by emp_cnt desc
'''

In [121]:
df = pd.read_sql_query(query_5_2, connection)
df

C:\Users\bogd_\AppData\Local\Temp\ipykernel_21948\3341481329.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)


areas_name  emp_cnt  vacancy_cnt
0                            Россия      410            0
1                         Казахстан      207            0
2                Московская область       75            0
3                Краснодарский край       19            0
4                          Беларусь       18            0
..                              ...      ...          ...
588                        Кувандык        0            0
589                        Куйбышев        0            0
590  Куйбышево (Ростовская область)        0            0
591                        Култаево        0            0
592                        Кульсары        0            0

[593 rows x 3 columns]

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [132]:
query_5_3 =f''' select e.name emp_name,
                count(distinct(v.area_id)) area_cnt
            from employers e 
            left join vacancies v on e.id=v.employer_id
            group by e.name
            order by area_cnt desc
           
'''

In [133]:
df = pd.read_sql_query(query_5_3, connection)
df

C:\Users\bogd_\AppData\Local\Temp\ipykernel_21948\2398953733.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)


emp_name  area_cnt
0                      Яндекс       181
1                  Ростелеком       152
2                  Спецремонт       116
3      Поляков Денис Иванович        88
4                   ООО ЕФИН         71
...                       ...       ...
23170                  СДЕЛКА         0
23171            Alandr Group         0
23172                     СДК         0
23173      Lemon Land Lombard         0
23174                  Махина         0

[23175 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [134]:
query_5_4 =f''' select count(e.name)
            from employers e 
            left join employers_industries e_i on e.id=e_i.employer_id
            where e_i.industry_id is null
           
'''

In [135]:
df = pd.read_sql_query(query_5_4, connection)
df

C:\Users\bogd_\AppData\Local\Temp\ipykernel_21948\3114411492.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_4, connection)


count
0   8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [147]:
query_5_5 =f''' select e.name
            from employers e 
            left join employers_industries e_i on e.id=e_i.employer_id
            group by e.name
            having count(e_i.industry_id) = 4
            order by e.name
            limit 3
           
'''

In [148]:
df = pd.read_sql_query(query_5_5, connection)
df

C:\Users\bogd_\AppData\Local\Temp\ipykernel_21948\1650419958.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


name
0  101 Интернет
1      21vek.by
2          2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [159]:
query_5_6 =f''' select count(e.name)
            from employers e 
            left join employers_industries e_i on e.id=e_i.employer_id
            left join industries i on e_i.industry_id=i.id
            where i.name = 'Разработка программного обеспечения'
           
            
           
'''

In [160]:
df = pd.read_sql_query(query_5_6, connection)
df

C:\Users\bogd_\AppData\Local\Temp\ipykernel_21948\1169039133.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6, connection)


count
0   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [163]:
import urllib.parse as req
path_common = req.urlparse('https://ru.wikipedia.org/wiki/Города-миллионеры_России#Список_городов-миллионеров')
url = path_common.scheme + '://' + req.quote(path_common.netloc) +  req.quote(path_common.path) + '#'+ req.quote(path_common.fragment)

In [ ]:
# текст запроса

In [ ]:
# результат запроса

***

In [ ]:
# выводы по анализу работодателей

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [175]:
query_6_1 =f''' select count(v.name)
            from vacancies v 
            where lower(v.name) like '%data%' or lower(v.name) like '%данн%'  
           
            
           
'''

In [176]:
df = pd.read_sql_query(query_6_1, connection)
df

C:\Users\bogd_\AppData\Local\Temp\ipykernel_21948\485020324.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_1, connection)


count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [190]:
query_6_2 =f''' SELECT 
    COUNT(DISTINCT id) vacancies_count,
    name vacancies_name
FROM VACANCIES 
WHERE 
    (
    lower(name) like '%data scientist%'  or
    lower(name) like '%data science%' or
    lower(name) like '%исследователь данных%'  or
    (name not like '%HTML%' and name like '%ML%')  or
    lower(name) like '%machine learning%'  or
    lower(name) like '%машинн%обучен%'
    )
    and
    (
    lower(name)  like '%junior%' or
    lower(experience) like '%нет опыта%' or
    lower(employment) like '%стажировка%'
    )
    GROUP BY name, id
            
           
'''

In [191]:
df = pd.read_sql_query(query_6_2, connection)
df

C:\Users\bogd_\AppData\Local\Temp\ipykernel_21948\3641650667.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


vacancies_count                                     vacancies_name
0                 1                 Research Engineer / Data Scientist
1                 1                            Junior+ CV/ML developer
2                 1          Senior Research Engineer / Data Scientist
3                 1  Исследователь в области машинного обучения и м...
4                 1                             Data Scientist, Junior
5                 1                             Data Scientist, Junior
6                 1                            Data Scientist (Junior)
7                 1            Data Scientist/ML engineer (Python/C++)
8                 1                              Ведущий AI/ML инженер
9                 1                    Data Scientist (junior+/middle)
10                1                                     Data Scientist
11                1  Программист Python/ Data Scientist (junior / m...
12                1                           Data Scientist (Junior+)
13                1                              Junior Data Scientist
14                1       Junior Data scientist/ Младший риск-аналитик
15                1  Стажер направления Машинное обучение и анализ ...
16                1              Python developer / Junior ML engineer
17                1                         Senior Data Scientist (AI)
18                1          Junior Research Engineer / Data Scientist
19                1             Junior Data Scientist for NLP (Python)
20                1                              Junior Data Scientist
21                1                              Data Scientist в Сбер
22                1        Data scientist (Junior/Middle) в консалтинг
23                1            Риск - Аналитик (Junior Data scientist)
24                1       Риск-аналитик (ML Engineer / Data Scientist)
25                1                       Intern/Junior Data Scientist
26                1                         Junior Data Scientist (VK)
27                1                       Junior/Middle Data Scientist
28                1  Junior Data Scientist (Валидация скоринговых м...
29                1  Junior Data Scientist (Валидация скоринговых м...
30                1  Junior Data Scientist (Валидация скоринговых м...
31                1  Junior Data Scientist (Валидация скоринговых м...
32                1  Junior Data Scientist (Валидация скоринговых м...
33                1  Junior Data Scientist (Валидация скоринговых м...
34                1  Junior Data Scientist (Валидация скоринговых м...
35                1  Junior Data Scientist (Валидация скоринговых м...
36                1  Junior Data Scientist (Валидация скоринговых м...
37                1                              Junior Data Scientist
38                1  Стажёр Data Scientist / Группа сервисов Антифр...
39                1                              Ведущий AI/ML инженер
40                1          Data Scientist (Computer Vision) / Junior
41                1  Data Scientist/Machine Learning Engineer (Deve...
42                1  Data Scientist/Machine Learning Engineer (Deve...
43                1  Data Scientist/Machine Learning Engineer (Deve...
44                1  Junior Data Scientist Управления модельных рис...
45                1          Data Scientist (Computer Vision) - Junior
46                1     Математик, Junior/Middle/Senior Data Scientist
47                1  Математик-аналитик (data scientist / data anal...
48                1                     Junior Data Scientist (Стажер)
49                1             Junior Data Scientist (Promo Forecast)
50                1  Junior Data Scientist (математик-аналитик) уда...

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [200]:
query_6_3=f"""
    select
        count(*)
    from
        public.vacancies
    where 
        (lower(name) like '%data scientist%' 
        or lower(name) like '%data science%'
        or lower(name) like '%исследователь данных%'
        or (name like '%ML%' and  lower(name) not like '%html%')
        or lower(name) like '%machine learning%'
        or lower(name) like '%машинн%обучен%')
        and (lower(key_skills)  like '%sql%' or lower(key_skills)  like '%postgres%')
"""

In [201]:
df = pd.read_sql_query(query_6_3, connection)
df

C:\Users\bogd_\AppData\Local\Temp\ipykernel_21948\1159587760.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


count
0    201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [203]:
query_6_4=f"""
    select
        count(*)
    from
        public.vacancies
    where 
        (lower(name) like '%data scientist%' 
        or lower(name) like '%data science%'
        or lower(name) like '%исследователь данных%'
        or (name like '%ML%' and  lower(name) not like '%html%')
        or lower(name) like '%machine learning%'
        or lower(name) like '%машинн%обучен%')
        and (lower(key_skills)  like '%python%')
"""

In [204]:
df = pd.read_sql_query(query_6_4, connection)
df

C:\Users\bogd_\AppData\Local\Temp\ipykernel_21948\972386973.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


count
0    351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [213]:
query_6_5=f"""
    select round(avg(length(key_skills) - length(replace(key_skills, CHR(9), ''))+1), 2) cnt_ks
    from
        public.vacancies
    where 
        (lower(name) like '%data scientist%' 
        or lower(name) like '%data science%'
        or lower(name) like '%исследователь данных%'
        or (name like '%ML%' and  lower(name) not like '%html%')
        or lower(name) like '%machine learning%'
        or lower(name) like '%машинн%обучен%')
        
"""

In [214]:
df = pd.read_sql_query(query_6_5, connection)
df

C:\Users\bogd_\AppData\Local\Temp\ipykernel_21948\390248827.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_5, connection)


cnt_ks
0    6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [219]:
query_vac_data_wage = f''' with table_1 as ((select
                                                v.experience as experience,
                                                avg(v.salary_from) as wage
                                            from 
                                                public.vacancies as v
                                            where 
                                                (v.name ilike any(array['%data scientist%', '%data science%', '%исследователь данных%', '%machine learning%', '%машинн%обучен%'])                     
                                                or v.name like '%ML%' 
                                                and v.name not ilike '%HTML%')
                                                and (v.salary_from is not null and v.salary_to is null)
                                            group by 
                                                v.experience)
                                            union
                                            (select
                                                v.experience as experience,
                                                avg(v.salary_to) as wage
                                            from 
                                                public.vacancies as v
                                            where 
                                                (v.name ilike any(array['%data scientist%', '%data science%', '%исследователь данных%', '%machine learning%', '%машинн%обучен%'])                     
                                                or v.name like '%ML%' 
                                                and v.name not ilike '%HTML%')
                                                and (v.salary_from is null and v.salary_to is not null)
                                            group by 
                                                v.experience)
                                            union
                                            (select
                                                v.experience as experience,
                                                avg((salary_from + v.salary_to)/2) as wage
                                            from 
                                                public.vacancies as v
                                            where 
                                                (v.name ilike any(array['%data scientist%', '%data science%', '%исследователь данных%', '%machine learning%', '%машинн%обучен%'])                     
                                                or v.name like '%ML%' 
                                                and v.name not ilike '%HTML%')
                                                and (v.salary_from is not null and v.salary_to is not null)
                                            group by 
                                                v.experience))
                                select
                                    t.experience as experience,
                                    round(avg(t.wage)) as avg_wage
                                from table_1 as t
                                group by t.experience
                                order by 2
                          '''

In [220]:
df = pd.read_sql_query(query_vac_data_wage, connection)
df

C:\Users\bogd_\AppData\Local\Temp\ipykernel_21948\130465342.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_vac_data_wage, connection)


experience  avg_wage
0           Нет опыта   72222.0
1  От 1 года до 3 лет  133311.0
2       От 3 до 6 лет  242400.0

***

Если коротко, то на рынке существует около 1700 вакансий, содержащих данн или data, однако  если добавлять условия (спец навыки), то их количества резко сокращается, это означает, что люди, работающие в сфере DS могут владеть определенным набором навыков, но при этом могут не подходить под другие вакансии той же сферы (DS). этот блок показал, что рынок вакансий разнообразен, на каждой вакансии ценяться разный набор навыков, однако неизменно одно - с увеличением опыта ЗП растет в прямой пропорции.